### PR TITLE
8342075: HttpClient: improve HTTP/2 flow control checks

### PR DESCRIPTION
**Pull request by dfuch**
Please find here a fix that improves flow control in the HTTP/2 implementation.

The change makes sure that flow control issues are reported to the server as FLOW_CONTROL_ERROR.
It also clarify how some system properties that allow to initialize flow control windows are handled, by documenting the full range of valid values (when applicable) and explaining what happens if the property points to a value that is out of range.

Bad flow control values in the SETTINGS frame will also cause a FLOW_CONTROL_ERROR to be reported.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Change requires CSR request [JDK-8342834](https://bugs.openjdk.org/browse/JDK-8342834) to be approved
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8342075](https://bugs.openjdk.org/browse/JDK-8342075): HttpClient: improve HTTP/2 flow control checks (**Bug** - P4)
 * [JDK-8342834](https://bugs.openjdk.org/browse/JDK-8342834): HttpClient: improve HTTP/2 flow control checks (**CSR**)

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21567/head:pull/21567` \
`$ git checkout pull/21567`

Update a local copy of the PR: \
`$ git checkout pull/21567` \
`$ git pull https://git.openjdk.org/jdk.git pull/21567/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21567`

View PR using the GUI difftool: \
`$ git pr show -t 21567`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21567.diff">https://git.openjdk.org/jdk/pull/21567.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21567#issuecomment-2420242055)

### CODE REVIEW

File: test/jdk/java/net/httpclient/http2/StreamFlowControlTest.java [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+            if (t instanceof FCHttp2TestExchange fct) {
+                fct.responseSent(query);
+            } else fail("Exchange is not %s but %s"
+                    .formatted(FCHttp2TestExchange.class.getName(), t.getClass().getName()));
+        }
+    }
+
+    // A custom Http2TestExchangeImpl that overrides sendResponseHeaders to
+    // allow headers to be sent with a number of CONTINUATION frames.
+    static class  FCHttp2TestExchange extends Http2TestExchangeImpl {
```
Comment (turbanoff) [inline test/jdk/java/net/httpclient/http2/StreamFlowControlTest.java:322]:
```suggestion
    static class FCHttp2TestExchange extends Http2TestExchangeImpl {
```

File: test/jdk/java/net/httpclient/http2/StreamFlowControlTest.java [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+        }
+
+    }
+
+    @Test(dataProvider = "variants")
+    void testAsync(String uri,
+                   boolean sameClient)
+    {
+        System.out.printf("%ntesting testAsync(%s, %s)%n", uri, sameClient);
+        ConcurrentHashMap<String, CompletableFuture<String>>  responseSent = new ConcurrentHashMap<>();
```
Comment (turbanoff) [inline test/jdk/java/net/httpclient/http2/StreamFlowControlTest.java:161]:
```suggestion
        ConcurrentHashMap<String, CompletableFuture<String>> responseSent = new ConcurrentHashMap<>();
```

File: test/jdk/java/net/httpclient/http2/ConnectionFlowControlTest.java [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+                { http2URI },
+                { https2URI },
+        };
+    }
+
+    @Test(dataProvider = "variants")
+    void test(String uri) throws Exception {
+        System.out.printf("%ntesting %s%n", uri);
+        ConcurrentHashMap<String, CompletableFuture<String>>  responseSent = new ConcurrentHashMap<>();
+        ConcurrentHashMap<String, HttpResponse<InputStream>>  responses = new ConcurrentHashMap<>();
```
Comment (turbanoff) [inline test/jdk/java/net/httpclient/http2/ConnectionFlowControlTest.java:116]:
```suggestion
        ConcurrentHashMap<String, CompletableFuture<String>> responseSent = new ConcurrentHashMap<>();
        ConcurrentHashMap<String, HttpResponse<InputStream>> responses = new ConcurrentHashMap<>();
```

File: test/jdk/java/net/httpclient/http2/ConnectionFlowControlTest.java [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+        @Override
+        public void handle(Http2TestExchange t) throws IOException {
+            String query = t.getRequestURI().getRawQuery();
+
+            try (InputStream is = t.getRequestBody();
+                 OutputStream os = t.getResponseBody()) {
+
+                byte[] bytes = is.readAllBytes();
+                System.out.println("Server " + t.getLocalAddress() + " received:\n"
+                        + t.getRequestURI() + ": " +  new String(bytes, StandardCharsets.UTF_8));
```
Comment (turbanoff) [inline test/jdk/java/net/httpclient/http2/ConnectionFlowControlTest.java:299]:
```suggestion
                        + t.getRequestURI() + ": " + new String(bytes, StandardCharsets.UTF_8));
```

File: test/jdk/java/net/httpclient/http2/ConnectionFlowControlTest.java [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+            } else {
+                fail("Exchange is not %s but %s"
+                        .formatted(FCHttp2TestExchange.class.getName(), t.getClass().getName()));
+            }
+        }
+    }
+
+    // A custom Http2TestExchangeImpl that overrides sendResponseHeaders to
+    // allow headers to be sent with a number of CONTINUATION frames.
+    static class  FCHttp2TestExchange extends Http2TestExchangeImpl {
```
Comment (turbanoff) [inline test/jdk/java/net/httpclient/http2/ConnectionFlowControlTest.java:345]:
```suggestion
    static class FCHttp2TestExchange extends Http2TestExchangeImpl {
```

File: src/java.net.http/share/classes/jdk/internal/net/http/Stream.java [hunk 6]
```diff
@@ -1426,12 +1450,18 @@ void cancel(IOException cause) {
...
         if (debug.on()) {
-            debug.log("cancelling exchange on stream %d due to protocol error: %s", streamid, cause.getMessage());
+            debug.log("cancelling exchange on stream %d due to protocol error [%s]: %s",
+                    streamid, ErrorFrame.stringForCode(code),
+                    cause.getMessage());
         }
         Log.logError("cancelling exchange on stream {0} due to protocol error: {1}\n", streamid, cause);
         // send a RESET frame and close the stream
-        cancelImpl(cause, ResetFrame.PROTOCOL_ERROR);
+        cancelImpl(cause,code);
```
Comment (turbanoff) [inline src/java.net.http/share/classes/jdk/internal/net/http/Stream.java:1464]:
```suggestion
        cancelImpl(cause, code);
```

File: src/java.net.http/share/classes/jdk/internal/net/http/WindowUpdateSender.java [hunk 7]
```diff
@@ -66,16 +81,116 @@ abstract class WindowUpdateSender {
...
+     * This method returns {@code true} if the bytes can be buffered
+     * without exceeding the flow control window, {@code false}
+     * if the flow control window is exceeded and corrective
+     * action (close/reset) has been taken.
+     * <p>
+     * When this method returns true, either {@link #processed(int)}
+     * or {@link #released(int)} must eventually be called to release
+     * the bytes from the flow control window.
+     *
+     * @implSpec
```
Comment (jaikiran) [inline src/java.net.http/share/classes/jdk/internal/net/http/WindowUpdateSender.java:107]: Is it intentional to use `@implSpec` here on an internal class. Or is it just to convey the intention?

File: src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java [hunk 8]
```diff
@@ -1080,6 +1081,32 @@ private String checkMaxOrphanedHeadersExceeded(HeaderFrame hf) {
...
+    // each frame that is dropped from the queue.
+    final void releaseUnconsumed(DataFrame df) {
+        windowUpdater.released(df.payloadLength());
+        dropDataFrame(df);
+    }
+
+    // This method is called when a DataFrame is dropped before/without
+    // having been added to any Stream::inputQ. In that case, the number
+    // of unprocessed bytes hasn't been incremented by the stream, and
+    // does not need to be decremented.
```
[the remark this answers is not in the document]
Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java:1109]: Maybe I should say "is called directly" ? The idea is that if the data is dropped before being added to the inputQ, then `dropDataFrame` is called directly. If the data is dropped after having been added to the inputQ, then `releaseUnconsumed` is called.

File: src/java.net.http/share/classes/module-info.java [hunk 9]
```diff
@@ -57,9 +57,11 @@
...
  * </li>
  * <li><p><b>{@systemProperty jdk.httpclient.connectionWindowSize}</b> (default: 2^26)<br>
- * The HTTP/2 client connection window size in bytes. The maximum size is 2^31-1. This value
- * cannot be smaller than the stream window size, which can be configured through the
- * {@code jdk.httpclient.windowsize} system property.
+ * The HTTP/2 client connection window size in bytes. Valid values are in the range
+ * [2^16-1, 2^31-1]. If an invalid value is provided, the default value is used.
+ * The implementation guarantees that the actual value will be no smaller than the stream
+ * window size, which can be configured through the {@code jdk.httpclient.windowsize}
+ * system property.
```
Comment (jaikiran) [inline src/java.net.http/share/classes/module-info.java:64]: Although these changes to the system property documentation are trivial, I think this would require a CSR, since these are "assertable" changes.

File: src/java.net.http/share/classes/jdk/internal/net/http/WindowUpdateSender.java [hunk 10]
```diff
@@ -66,16 +81,116 @@ abstract class WindowUpdateSender {
...
+    boolean canBufferUnprocessedBytes(int len) {
+        return !checkWindowSizeExceeded(unprocessed.addAndGet(len));
+    }
+
+    // adds the provided amount to the amount of already
+    // received and processed bytes and checks whether the
+    // flow control window is exceeded. If so, take
+    // corrective actions and return true.
+    private boolean checkWindowSizeExceeded(int len) {
+        int rcv = Math.addExact(received.get(), len);
```
[the remark this answers is not in the document]
Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/WindowUpdateSender.java:127]: Good point. Maybe I should use a long to accumulate the two integers,

File: src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (jaikiran) [reply src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java:1109]: > Maybe I should say "is called directly" ? 

I think that would still be unclear since "directly" is contextual and depends on who is calling it. So we would then have to clarify what a direct call is.

Perhaps we should reword it to say that the dropDataFrame method gets called in two different ways. One when the data is dropped before adding to the inputQ and the other when the data is dropped through releaseUnconsumed.

**Review by jaikiran — APPROVED** [review]
Thank you for the updates Daniel. This looks good to me.

**Comment by dfuch** [comment]
I see the CSR got approved. I'm running a last batch of tier2 tests and will integrate if they come back successful.
